File: src/apple-tv/flow.ts
import { Proto } from '@basmilius/apple-sdk';
import { Shortcuts } from '@basmilius/homey-common';
import type { AppleApp, AppleTVDevice } from '../types';
import { repeatModeToProto } from '../utils';
import type Homey from 'homey';

export default class AppleTVFlow extends Shortcuts<AppleApp> {
    register(): void {
        this.#registerLaunchApp();
        this.#registerLaunchUrl();
        this.#registerNowPlayingAppBecomes();
        this.#registerNowPlayingAppIs();
        this.#registerRemote();
        this.#registerSetPosition();
        this.#registerSetRepeat();
        this.#registerSetShuffle();
        this.#registerSkipBackward();
        this.#registerSkipForward();
        this.#registerSwitchAccount();
    }

    readonly #appAutocompleteListener = async (query: string, {device}: { device: AppleTVDevice }): Promise<Homey.FlowCard.ArgumentAutocompleteResults> => {
        const launchableApps = await device.sdk.apps?.list() ?? [];
        const lowerQuery = query.trim().toLowerCase();

        return launchableApps
            .filter(app => lowerQuery.length === 0 || app.name.toLowerCase().includes(lowerQuery))
            .map(app => ({
                id: app.bundleId,
                name: app.name,
                description: app.bundleId
            }))
            .sort((a, b) => a.name.localeCompare(b.name));
    };

    async triggerCompanionLinkFailed(device: AppleTVDevice): Promise<void> {
        try {
            const triggerCard = this.flow.getDeviceTriggerCard('appletv_companion_link_failed');
            await triggerCard.trigger(device);
        } catch (err) {
            this.log(device.name, 'Failed to trigger companion link failed card.', err);
        }
    }

    async triggerArtworkUrlUpdated(device: AppleTVDevice, localUrl: string, cloudUrl: string): Promise<void> {
        try {
            const triggerCard = this.flow.getDeviceTriggerCard('appletv_artwork_url_updated');

            await triggerCard.trigger(device, {
                localUrl,
                cloudUrl
            });
        } catch (err) {
            this.log(device.name, 'Failed to trigger artwork url updated card.', err);
        }
    }

    async triggerNowPlayingAppChanges(device: AppleTVDevice, bundleIdentifier: string, displayName: string): Promise<void> {
        try {
            const triggerCard = this.flow.getDeviceTriggerCard('appletv_now_playing_app_changes');

            await triggerCard.trigger(device, {
                bundleIdentifier,
                displayName
            });
        } catch (err) {
            this.log(device.name, 'Failed to trigger now playing app changes card.', err);
        }
    }

    async triggerNowPlayingAppBecomes(device: AppleTVDevice, bundleIdentifier: string, displayName: string): Promise<void> {
        try {
            const triggerCard = this.flow.getDeviceTriggerCard('appletv_now_playing_app_becomes');

            await triggerCard.trigger(device, {}, {
                bundleId: bundleIdentifier,
                displayName
            });
        } catch (err) {
            this.log(device.name, 'Failed to trigger now playing app becomes card.', err);
        }
    }

    #registerLaunchApp(): void {
        const launchApp = this.flow.getActionCard('appletv_launch_app');

        type RunArguments = {
            readonly app: AppleApp;
            readonly device: AppleTVDevice;
        };

        launchApp.registerRunListener(async ({app, device}: RunArguments) => {
            await device.sdk.apps?.launch(app.id);
        });

        launchApp.registerArgumentAutocompleteListener('app', this.#appAutocompleteListener);
    }

    #registerNowPlayingAppBecomes(): void {
        const card = this.flow.getDeviceTriggerCard('appletv_now_playing_app_becomes');

        type RunArguments = {
            readonly app: AppleApp;
        };

        type State = {
            readonly bundleId: string;
        };

        card.registerRunListener(async (args: RunArguments, state: State) => {
            return args.app.id === state.bundleId;
        });

        card.registerArgumentAutocompleteListener('app', this.#appAutocompleteListener);
    }

    #registerNowPlayingAppIs(): void {
        const card = this.flow.getConditionCard('appletv_now_playing_app_is');

        type RunArguments = {
            readonly app: AppleApp;
            readonly device: AppleTVDevice;
        };

        card.registerRunListener(async ({app, device}: RunArguments) => {
            return device.currentNowPlayingBundleId === app.id;
        });

        card.registerArgumentAutocompleteListener('app', this.#appAutocompleteListener);
    }

    #registerLaunchUrl(): void {
        const launchUrl = this.flow.getActionCard('appletv_launch_url');

        type RunArguments = {
            readonly device: AppleTVDevice;
            readonly url: string;
        }

        launchUrl.registerRunListener(async ({device, url}: RunArguments) => {
            await device.sdk.apps?.openUrl(url);
        });
    }

    #registerRemote(): void {
        const remote = this.flow.getActionCard('appletv_remote');

        type RunArguments = {
            readonly device: AppleTVDevice;
            readonly command: string;
        }

        remote.registerRunListener(async ({device, command}: RunArguments) => {
            const sdk = device.sdk;

            const commands: Record<string, () => Promise<void>> = {
                up: () => sdk.remote.up(),
                down: () => sdk.remote.down(),
                left: () => sdk.remote.left(),
                right: () => sdk.remote.right(),
                select: () => sdk.remote.select(),
                menu: () => sdk.remote.menu(),
                home: () => sdk.remote.home(),
                play: () => sdk.remote.play(),
                pause: () => sdk.remote.pause(),
                playPause: () => sdk.remote.playPause(),
                next: () => sdk.remote.next(),
                previous: () => sdk.remote.previous(),
                volumeUp: () => sdk.volume.up(),
                volumeDown: () => sdk.volume.down(),
                mute: () => sdk.remote.mute(),
                stop: () => sdk.remote.stop(),
                screensaver: () => {
                    if (!sdk.companionLink) { throw new Error('Companion Link is not available.'); }
                    return sdk.companionLink.pressButton('Screensaver');
                },
                topMenu: () => sdk.remote.topMenu(),
                homeHold: () => {
                    if (!sdk.companionLink) { throw new Error('Companion Link is not available.'); }
                    return sdk.companionLink.pressButton('Home', 'Hold');
                },
                doubleTapSelect: () => {
                    if (!sdk.companionLink) { throw new Error('Companion Link is not available.'); }
                    return sdk.companionLink.pressButton('Select', 'DoubleTap');
                },
                channelUp: () => sdk.remote.channelUp(),
                channelDown: () => sdk.remote.channelDown(),
                siri: () => {
                    if (!sdk.companionLink) { throw new Error('Companion Link is not available.'); }
                    return sdk.companionLink.pressButton('Siri');
                },
                wake: () => sdk.remote.wake(),
                suspend: () => sdk.remote.suspend()
            };

            const fn = commands[command];

            if (!fn) {
                throw new Error(`Unknown remote command: ${command}`);
            }

            await fn();
        });
    }

    #registerSetPosition(): void {
        const card = this.flow.getActionCard('appletv_set_position');

        type RunArguments = {
            readonly device: AppleTVDevice;
            readonly position: number;
        };

        card.registerRunListener(async ({device, position}: RunArguments) => {
            await device.sdk.playback.seekTo(Math.floor(position));
        });
    }

    #registerSetRepeat(): void {
        const card = this.flow.getActionCard('appletv_set_repeat');

        type RunArguments = {
            readonly device: AppleTVDevice;
            readonly mode: string;
        };

        card.registerRunListener(async ({device, mode}: RunArguments) => {
            await device.sdk.playback.setRepeatMode(repeatModeToProto[mode] ?? Proto.RepeatMode_Enum.Off);
        });
    }

    #registerSetShuffle(): void {
        const card = this.flow.getActionCard('appletv_set_shuffle');

        type RunArguments = {
            readonly device: AppleTVDevice;
            readonly shuffle: string;
        };

        card.registerRunListener(async ({device, shuffle}: RunArguments) => {
            const mode = shuffle === 'true'
                ? Proto.ShuffleMode_Enum.Songs
                : Proto.ShuffleMode_Enum.Off;
            await device.sdk.playback.setShuffleMode(mode);
        });
    }

    #registerSkipBackward(): void {
        const card = this.flow.getActionCard('appletv_skip_backward');

        type RunArguments = {
            readonly device: AppleTVDevice;
            readonly seconds: number;
        };

        card.registerRunListener(async ({device, seconds}: RunArguments) => {
            await device.sdk.playback.skipBackward(Math.floor(seconds));
        });
    }

    #registerSkipForward(): void {
        const card = this.flow.getActionCard('appletv_skip_forward');

        type RunArguments = {
            readonly device: AppleTVDevice;
            readonly seconds: number;
        };

        card.registerRunListener(async ({device, seconds}: RunArguments) => {
            await device.sdk.playback.skipForward(Math.floor(seconds));
        });
    }

    #registerSwitchAccount(): void {
        const switchAccount = this.flow.getActionCard('appletv_switch_account');

        type AutocompleteArguments = {
            readonly device: AppleTVDevice;
        };

        type RunArguments = {
            readonly account: {
                readonly id: string;
            };
            readonly device: AppleTVDevice;
        };

        switchAccount.registerRunListener(async ({device, account}: RunArguments) => {
            await device.sdk.accounts?.switch(account.id);
        });

        switchAccount.registerArgumentAutocompleteListener('account', async (query: string, {device}: AutocompleteArguments): Promise<Homey.FlowCard.ArgumentAutocompleteResults> => {
            const userAccounts = await device.sdk.accounts?.list() ?? [];
            const lowerQuery = query.trim().toLowerCase();

            return userAccounts
                .filter(account => lowerQuery.length === 0 || account.name.toLowerCase().includes(lowerQuery))
                .map(account => ({
                    id: account.accountId,
                    name: account.name
                }))
                .sort((a, b) => a.name.localeCompare(b.name));
        });
    }
}

File: src/apple_tv/flow.py
from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING, Any

from pyatv.const import InputAction, RepeatState, ShuffleState


import pyatv.interface as pyatv_interface

if TYPE_CHECKING:
    from ..app import AppleApp
    from .device import AppleTVDevice

# Maps flow command names to remote_control method names.
_REPEAT_MAP: dict[str, RepeatState] = {
    'off': RepeatState.Off,
    'one': RepeatState.Track,
    'all': RepeatState.All,
}

_REMOTE_COMMAND_METHODS: dict[str, str] = {
    'up': 'up',
    'down': 'down',
    'left': 'left',
    'right': 'right',
    'select': 'select',
    'menu': 'menu',
    'home': 'home',
    'play': 'play',
    'pause': 'pause',
    'playPause': 'play_pause',
    'next': 'next',
    'previous': 'previous',
    'volumeUp': 'volume_up',
    'volumeDown': 'volume_down',
}


def _require_atv(args: dict[str, Any], label: str) -> tuple[AppleTVDevice, pyatv_interface.AppleTV]:
    """Extract device and active pyatv connection from flow args, or raise."""
    device: AppleTVDevice = args['device']
    atv = device.atv
    if atv is None:
        raise RuntimeError(f'{label} "{device.get_name()}" is not connected')
    return device, atv


class AppleTVFlow:
    """Registers and manages Apple TV flow cards."""

    def __init__(self, app: AppleApp) -> None:
        self._app = app

    def register(self) -> None:
        self._register_launch_app()
        self._register_launch_url()
        self._register_remote()
        self._register_set_position()
        self._register_skip_forward()
        self._register_skip_backward()
        self._register_set_repeat()
        self._register_set_shuffle()
        self._register_switch_account()

    async def trigger_artwork_url_updated(
        self, device: AppleTVDevice, local_url: str, cloud_url: str
    ) -> None:
        try:
            card = self._app.homey.flow.get_device_trigger_card(
                'appletv_artwork_url_updated'
            )
            await card.trigger(device, {'localUrl': local_url, 'cloudUrl': cloud_url})
        except asyncio.CancelledError:
            raise
        except Exception as err:
            self._app.log(device.get_name(), 'Failed to trigger artwork url updated card.', err)

    async def trigger_now_playing_app_changes(
        self,
        device: AppleTVDevice,
        bundle_identifier: str,
        display_name: str,
    ) -> None:
        try:
            card = self._app.homey.flow.get_device_trigger_card(
                'appletv_now_playing_app_changes'
            )
            await card.trigger(
                device,
                {'bundleIdentifier': bundle_identifier, 'displayName': display_name},
            )
        except asyncio.CancelledError:
            raise
        except Exception as err:
            self._app.log(
                device.get_name(), 'Failed to trigger now playing app changes card.', err
            )

    # ------------------------------------------------------------------
    # Action card registrations
    # ------------------------------------------------------------------

    def _register_launch_app(self) -> None:
        card = self._app.homey.flow.get_action_card('appletv_launch_app')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            device, atv = _require_atv(args, 'Apple TV')
            app_arg: Any = args.get('app')
            if not isinstance(app_arg, dict) or 'id' not in app_arg:
                raise ValueError('No app selected')
            await atv.apps.launch_app(app_arg['id'])

        async def autocomplete(query: str, **kwargs: Any) -> list[dict]:
            device: AppleTVDevice = kwargs['device']
            atv = device.atv
            if atv is None:
                return []
            try:
                app_list = await atv.apps.app_list()
            except Exception:
                return []
            results = [
                {
                    'id': a.identifier,
                    'name': a.name,
                    'description': a.identifier,
                }
                for a in app_list
                if not query.strip() or query.lower() in (a.name or '').lower()
            ]
            return sorted(results, key=lambda x: x['name'] or '')

        card.register_run_listener(run)
        card.get_argument('app').register_autocomplete_listener(autocomplete)

    def _register_launch_url(self) -> None:
        card = self._app.homey.flow.get_action_card('appletv_launch_url')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'Apple TV')
            url: str = args['url']
            await atv.apps.launch_app(url)

        card.register_run_listener(run)

    def _register_remote(self) -> None:
        card = self._app.homey.flow.get_action_card('appletv_remote')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            device, atv = _require_atv(args, 'Apple TV')
            command: str = args['command']

            rc = atv.remote_control

            method_name = _REMOTE_COMMAND_METHODS.get(command)
            if method_name is not None:
                await getattr(rc, method_name)()
                return

            # Commands that use different interfaces or methods.
            if command == 'stop':
                await rc.stop()
            elif command == 'screensaver':
                await rc.screensaver()
            elif command == 'topMenu':
                await rc.top_menu()
            elif command == 'homeHold':
                await rc.home_hold()
            elif command == 'doubleTapSelect':
                await rc.select(action=InputAction.DoubleTap)
            elif command == 'channelUp':
                await rc.channel_up()
            elif command == 'channelDown':
                await rc.channel_down()
            elif command == 'siri':
                # NOTE: Siri uses internal pyatv API (rc.api / HidCommand.Siri) as
                # there is no public/stable Siri interface in pyatv as of 0.17.x.
                # This may break with future pyatv updates.
                if hasattr(rc, 'api'):
                    from pyatv.protocols.companion.api import HidCommand
                    await rc.api.hid_command(True, HidCommand.Siri)
                    await rc.api.hid_command(False, HidCommand.Siri)
                else:
                    self._app.log(device.get_name(), 'Siri requires Companion Link protocol')
            elif command == 'mute':
                await device.toggle_mute()
            elif command == 'wake':
                await atv.power.turn_on()
            elif command == 'suspend':
                await atv.power.turn_off()
            else:
                raise ValueError(f'Unsupported remote command: {command}')

        card.register_run_listener(run)

    def _register_set_position(self) -> None:
        card = self._app.homey.flow.get_action_card('appletv_set_position')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'Apple TV')
            await atv.remote_control.set_position(int(float(args['position'])))

        card.register_run_listener(run)

    def _register_skip_forward(self) -> None:
        card = self._app.homey.flow.get_action_card('appletv_skip_forward')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'Apple TV')
            await atv.remote_control.skip_forward(int(float(args['seconds'])))

        card.register_run_listener(run)

    def _register_skip_backward(self) -> None:
        card = self._app.homey.flow.get_action_card('appletv_skip_backward')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'Apple TV')
            await atv.remote_control.skip_backward(int(float(args['seconds'])))

        card.register_run_listener(run)

    def _register_set_repeat(self) -> None:
        card = self._app.homey.flow.get_action_card('appletv_set_repeat')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'Apple TV')
            mode = args['mode']
            if mode not in _REPEAT_MAP:
                raise ValueError(f'Unsupported repeat mode: {mode!r}')
            await atv.remote_control.set_repeat(_REPEAT_MAP[mode])

        card.register_run_listener(run)

    def _register_set_shuffle(self) -> None:
        card = self._app.homey.flow.get_action_card('appletv_set_shuffle')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'Apple TV')
            shuffle = str(args['shuffle']).lower() == 'true'
            state = ShuffleState.Songs if shuffle else ShuffleState.Off
            await atv.remote_control.set_shuffle(state)

        card.register_run_listener(run)

    def _register_switch_account(self) -> None:
        card = self._app.homey.flow.get_action_card('appletv_switch_account')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'Apple TV')
            account: Any = args.get('account')
            if not isinstance(account, dict) or 'id' not in account:
                raise ValueError('No account selected')

            await atv.user_accounts.switch_account(account['id'])

        async def autocomplete(query: str, **kwargs: Any) -> list[dict]:
            device: AppleTVDevice = kwargs['device']
            atv = device.atv
            if atv is None:
                return []
            try:
                accounts = await atv.user_accounts.account_list()
            except Exception:
                return []
            results = [
                {'id': a.identifier, 'name': a.name}
                for a in accounts
                if not query.strip() or query.lower() in (a.name or '').lower()
            ]
            return sorted(results, key=lambda x: x['name'] or '')

        card.register_run_listener(run)
        card.get_argument('account').register_autocomplete_listener(autocomplete)
